x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java

```diff
@@ -245,18 +248,24 @@ public void clusterStateProcessed(ClusterState oldState, ClusterState newState)
 
     private static ClusterState update(ClusterState currentState, TrainedModelAssignmentMetadata.Builder modelAssignments) {
         if (modelAssignments.isChanged()) {
-            return ClusterState.builder(currentState)
-                .metadata(
-                    Metadata.builder(currentState.metadata())
-                        .putCustom(TrainedModelAssignmentMetadata.NAME, modelAssignments.build())
-                        .removeCustom(TrainedModelAssignmentMetadata.DEPRECATED_NAME)
-                )
-                .build();
+            return forceUpdate(currentState, modelAssignments);
         } else {
             return currentState;
         }
     }
 
+    private static ClusterState forceUpdate(ClusterState currentState, TrainedModelAssignmentMetadata.Builder modelAssignments) {
+        Metadata.Builder metadata = Metadata.builder(currentState.metadata());
+        if (currentState.getNodes().getMinNodeVersion().onOrAfter(RENAME_ALLOCATION_TO_ASSIGNMENT_VERSION)) {
+            metadata.putCustom(TrainedModelAssignmentMetadata.NAME, modelAssignments.build())
+                .removeCustom(TrainedModelAssignmentMetadata.DEPRECATED_NAME);
+        } else {
+            metadata.putCustom(TrainedModelAssignmentMetadata.DEPRECATED_NAME, modelAssignments.build())
+                .removeCustom(TrainedModelAssignmentMetadata.NAME);
```

Comment:
Is this removal necessary?  Surely it's only needed if we've already released software that prematurely upgrades?